Add "ManiaModAccelerate"

cdwcgt opened the pull request:
This is a mania mod, scroll speed will change with current combo.
If you miss, scroll speed will reduce to minspeed which can set in mod settings

~~- [ ] Might need a better icon, I don't think this one is good enough, but I'm out of ideas~~ no icon for now
- [x] The function may not be perfect, suggestions are welcome, or we may have more than one function

Now combo and scroll speed are positively correlated

https://user-images.githubusercontent.com/34775378/208305548-5d2dc0f1-e1bb-45ed-945a-150ca5c1b889.mp4

cdwcgt committed:
add ManiaModSlowDown

cdwcgt committed:
Merge branch 'master' into mod-slowdown

cdwcgt committed:
dynamic Max Min value

cdwcgt committed:
`ManiaModSlowDown` to `ManiaModScrollUp`

Co-Authored-By: Alleyne-Angel <67787327+alleyne-angel@users.noreply.github.com>

cdwcgt committed:
Add Test for scrollup

cdwcgt committed:
add Description

Co-Authored-By: Alleyne-Angel <67787327+Alleyne-Angel@users.noreply.github.com>

mk56-spn commented:
DK about the name, since upscroll (direction)  is a thing and its the first thing this mod made me think its about

cdwcgt commented:
> DK about the name, since upscroll (direction) is a thing and its the first thing this mod made me think its about

I originally thought about speedup, but this might conflict with windup
So I chose this name (from my friend, I'm not good at naming names), but if there is a better name, I will use it

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModScrollUp.cs:
```diff
@@ -0,0 +1,100 @@
+{
+    public class ManiaModScrollUp : Mod, IApplicableToDrawableRuleset<ManiaHitObject>, IApplicableToScoreProcessor, IApplicableToPlayer
+    {
+        public override string Name => "Scroll up";
```
name needs to change as it is confusing. "Accelerate" maybe

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModScrollUp.cs:
```diff
@@ -0,0 +1,100 @@
+        public override string Acronym => "SU";
+        public override LocalisableString Description => @"Key will become faster..., until you miss";
+        public override double ScoreMultiplier => 1;
+        public override IconUsage? Icon => FontAwesome.Solid.AngleDoubleUp;
```
remove icon for now. can't find it now but it was decided in one of these that new mods should not be choosing their own icons

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModScrollUp.cs:
```diff
@@ -0,0 +1,100 @@
+
+                double speed = MinScoreSpeed.Value - (MinScoreSpeed.Value - MaxScoreSpeed.Value) * Math.Log(s.NewValue + 1, MaxComboCount.Value + 1);
+
+                scoreProcessor.TransformBindableTo(scrollTime, speed, 500, Easing.OutQuint);
```
inefficient, this is going to likely cause tens to hundreds of transforms to be running at any given time. i'm also not even sure that this is going to work correctly with multiple overlapping notes

cdwcgt replied:
I refer to this https://github.com/ppy/osu/blob/b35e796d75153b53c1e450f407a48032aef1eddf/osu.Game/Rulesets/Mods/ModMuted.cs#L89
Because this value needs use transform instead of set directly, otherwise it will cause weird visual effects, and I didn't find a better solution either.
If there is a better way please let me know

bdach replied:
see mods that implement `IUpdatableByPlayfield`

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModScrollUp.cs:
```diff
@@ -0,0 +1,100 @@
+
+        public ManiaModScrollUp()
+        {
+            MinScoreSpeed.BindValueChanged(s =>
+            {
+                MaxScoreSpeed.MaxValue = s.NewValue;
+            });
+
+            MaxScoreSpeed.BindValueChanged(s =>
+            {
+                MinScoreSpeed.MinValue = s.NewValue;
+            });
```
so touching either value sets the other to the same value too? very weird. compare `ModWind{Up,Down}` for how this should actually be done

cdwcgt committed:
Improved value conflicts handle

cdwcgt committed:
remove icon

Flamiii commented:
"Accelerate" maybe?

Kori3 commented:
I feel like this mod fits the "Fun" category better rather than the conversion one

cdwcgt committed:
some logic fix

bug

Co-Authored-By: Alleyne-Angel <67787327+alleyne-angel@users.noreply.github.com>

cdwcgt committed:
move to `Fun` mod type

cdwcgt committed:
rename to `Accelerate`

cdwcgt committed:
use `IUpdatableByPlayfield`

cdwcgt committed:
Merge branch 'master' into mod-slowdown

cdwcgt committed:
change setting slider to scroll speed

cdwcgt committed:
ElapsedFrameTime may nagetive

Unexpected things will happen

cdwcgt committed:
Change calculation method for ManiaModAccelerate

Now combo and scroll speed are positively correlated

cdwcgt committed:
Make smoothTimeRange overridable by enabled scrollTime

cdwcgt committed:
make the test more reliable

Now it can test if the mod work, not just if it can load

cdwcgt committed:
Incompatible with ManiaModConstantSpeed

cdwcgt committed:
Merge branch 'master' into mod-slowdown

cdwcgt committed:
rename `ScrollSpeed` to `CustomSmoothTimeRange`

ScrollSpeed has been used and will cause mislead

cdwcgt committed:
Set the default range to 18-30

It should be within the acceptable range of normal players.

Co-Authored-By: Hidden is fun <54774053+hidden-is-fun@users.noreply.github.com>

cdwcgt committed:
add BaseComboCount

if combo is lower then BaseComboCount, speed will not change.

cdwcgt committed:
fix typo

score => scroll

cdwcgt committed:
Merge branch 'master' into mod-slowdown

cdwcgt committed:
fix test

cdwcgt committed:
fix conflict with AC (Accuracy Challenge)

but still, I don't think this abbreviation is appropriate

cdwcgt committed:
Merge branch 'master' into mod-slowdown

cdwcgt committed:
Merge branch 'master' into mod-slowdown

cdwcgt committed:
Add interface IManiaAdjustScrollSpeed

use interface to adjust scroll speed is better than add a separate member in drawable mania ruleset

cdwcgt committed:
some rework by using IManiaAdjustScrollSpeed interface

cdwcgt committed:
remove incompatible between `Accelerate` and `Constant Speed`

I might be stupid

cdwcgt committed:
fix test
